### ISSUE TITLE
`functional/immutable-data` working differently for TS and JS files

### ISSUE THREAD

**Issue by lhrinaldi** [jonaskello/eslint-plugin-functional#222]
# Bug Report

I'm not sure if this is a bug or it's expected, but `functional/immutable-data` works differently for JS and TS files.

## Expected behavior

I'd expect that the errors would be the same both for JS and TS files.

### Report on JS

![image](https://user-images.githubusercontent.com/1268808/125984091-e349c120-73d5-4b2c-af9b-597095e345a8.png)

### Report on TS

![image](https://user-images.githubusercontent.com/1268808/125984120-2e0d9c1a-eae8-4c9f-8f7d-945552ba7457.png)

## Steps to reproduce

- `npx create-react-app my-app --template=typescript`
- Use this config:
```json
"eslintConfig": {
    "extends": [
      "react-app"
    ],
    "plugins": [
      "functional"
    ],
    "rules": {
      "functional/immutable-data": "error"
    }
  }
```
- Paste this example to `App.tsx`
```js
const arr = [0, 1, 2];

arr[0] = 4; // <- Modifying an array is not allowed.
arr.length = 1; // <- Modifying an array is not allowed.
arr.push(3); // <- Modifying an array is not allowed.
arr.sort();
```
- Change the file extension from `.tsx` to `.jsx/.js`.

## Proposed changes

I want to know if this is actually really a bug and if it's I'll try to find a fix.

**Comment by lhrinaldi** [comment]
From debugging I found that https://github.com/jonaskello/eslint-plugin-functional/blob/master/src/rules/immutable-data.ts#L291 this function (`isArrayType`) doesn't returns true for TS.

**Comment by lhrinaldi** [comment]
And this is the output of `node.callee.object`.

```
{
  type: 'Identifier',
  name: 'arr',
  range: [ 149, 152 ],
  loc: { start: { line: 8, column: 2 }, end: { line: 8, column: 5 } },
  parent: {
    type: 'MemberExpression',
    object: [Circular],
    property: {
      type: 'Identifier',
      name: 'sort',
      range: [Array],
      loc: [Object],
      parent: [Circular]
    },
    computed: false,
    optional: false,
    range: [ 149, 157 ],
    loc: { start: [Object], end: [Object] },
    parent: {
      type: 'CallExpression',
      callee: [Circular],
      arguments: [],
      optional: false,
      range: [Array],
      loc: [Object],
      parent: [Object]
    }
  }
}
```

**Comment by RebeccaStevens** [comment]
JavaScript will assume types by default (see [`assumeTypes`](https://github.com/jonaskello/eslint-plugin-functional/blob/master/docs/rules/immutable-data.md#assumetypes)) while TypeScript relies on using TypeScript's type engine.

I believe the issue you're running into is that you have not enabled TypeScript's type engine in ESLint.
Adjust your ESLint config by adding this:

```json
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "project": "./tsconfig.json"
  }
}
```

See [with TypeScript](https://github.com/jonaskello/eslint-plugin-functional#with-typescript) for details.

**Comment by lhrinaldi** [comment]
That was it, thanks :) I thought CRA would already use a TS parser for TS projects.
